macros/validation/validate_grain.sql

```diff
@@ -0,0 +1,13 @@
+{% macro validate_grain(grain, full_set) %}
+
+    {# We loop through the full set here to ensure that the provided grain works for all metrics
+    returned or used, not just those listed #}
+
+    {% for metric_name in full_set %}
+        {% set metric_relation = metric(metric_name)%}
+        {% if grain not in metric_relation.time_grains%}
+            {%- do exceptions.raise_compiler_error("The metric " ~ metric_name ~ " does not have the provided time grain " ~ grain ~ "defined in the metric definition.") %}
```

Comment:
I'm thinking about when this would be triggered, and I guess it's most likely to be when an upstream metric is missing it. So it might be weird to request `metrics.calculate(revenue_per_user)` and get an error message talking about the `users` metric or something. 

Something like "Can't calculate metric by [grain] because the metric [metric_name] does not allow that grain"? If you can get hold of the data, it feels like the best experience would be to do:

if requested_metric != misconfigured_metric:
Can't calculate [requested_metric] by [grain] because it requires [misconfigured_metric], which does not support [grain]-level data.

if requested_metric == misconfigured_metric:
Can't calculate [requested_metric] by [grain] because it does not support [grain]-level data.